core/src/thread_parker/wasm.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2016 Amanieu d'Antras
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! A simple spin lock based thread parker. Used on platforms without better
+//! parking facilities available.
+
+use core::{
+    arch::wasm32,
+    mem,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+use std::{convert::TryFrom, thread, time::Instant};
+
+// Helper type for putting a thread to sleep until some other thread wakes it up
+pub struct ThreadParker {
+    parked: AtomicUsize,
```

Comment:
I'm not sure why this is not an `AtomicI32`. I just imitated what `libstd/sys/wasm/mutex_atomic.rs` does. The wait operation requires an `*mut i32`, and the `ptr()` method asserts that the sizes are equal. Maybe simply because this was written before `AtomicI32` was stabilized? If that's the case I guess this can be changed now.